File: src/editor/util/Containment.java
package editor.util;

import java.util.Collection;
import java.util.function.BiPredicate;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * This enumeration represents a way that elements from one #Collection can be contained in another.
 *
 * @author Alec Roelke
 */
public enum Containment implements BiPredicate<Collection<?>, Collection<?>>
{
    /**
     * The first collection contains all of the elements in the second collection.
     * Opposite of {@link #CONTAINS_NOT_ALL_OF}.
     */
    CONTAINS_ALL_OF,
    /**
     * The first collection contains any element in the second collection.  A collection always contains
     * an empty collection.
     * Opposite of {@link #CONTAINS_NONE_OF}.
     */
    CONTAINS_ANY_OF,
    /**
     * The first collection contains exactly the same elements as the second collection.
     * Opposite of {@link #CONTAINS_NOT_EXACTLY}.
     */
    CONTAINS_EXACTLY,
    /**
     * The first collection contains no elements from the second collection.
     * Opposite of {@link #CONTAINS_ANY_OF}.
     */
    CONTAINS_NONE_OF,
    /**
     * The first collection contains some, but not all, of the elements in the second collection.
     * Opposite if {@link #CONTAINS_ALL_OF}.
     */
    CONTAINS_NOT_ALL_OF,
    /**
     * The first collection contains some or none of the elements as the second collection, but is not
     * exactly the same.  Opposite of {@link #CONTAINS_EXACTLY}.
     */
    CONTAINS_NOT_EXACTLY;

    /**
     * @param contain String representation of the desired Containment
     * @return The Containment corresponding to the given String.
     */
    public static Containment parseContainment(String contain)
    {
        return Containment.valueOf(contain.toUpperCase().replace(' ', '_'));
    }

    @Override
    public boolean test(Collection<?> a, Collection<?> b)
    {
        return switch (this) {
            case CONTAINS_ANY_OF -> b.isEmpty() || b.stream().anyMatch(a::contains);
            case CONTAINS_NONE_OF -> b.stream().noneMatch(a::contains);
            case CONTAINS_ALL_OF -> a.containsAll(b);
            case CONTAINS_NOT_ALL_OF -> CONTAINS_ANY_OF.test(a, b) && !a.containsAll(b);
            case CONTAINS_EXACTLY -> a.stream().collect(Collectors.groupingBy(Function.identity(), Collectors.counting())).equals(
                                     b.stream().collect(Collectors.groupingBy(Function.identity(), Collectors.counting())));
            case CONTAINS_NOT_EXACTLY -> a.stream().anyMatch((o) -> !b.contains(o)) || b.stream().anyMatch((o) -> !a.contains(o));
        };
    }

    /**
     * {@inheritDoc}
     * The String representation of a Containment is its name in lower case with
     * _ replaced with a space.
     */
    @Override
    public String toString()
    {
        return super.toString().replace("_", " ").toLowerCase();
    }
}

File: src/main/scala/editor/util/Containment.scala
package editor.util

/**
 * Global data about [[Containment]].
 * @author Alec Roelke
 */
object Containment {
  /**
   * Parse a [[Containment]] from a string using its string representation rather than its name.
   * 
   * @param s string to parse
   * @return the [[Containment]] whose representation matches the given string, ignoring case, or None if there isn't one
   */
  def parse(s: String) = Containment.values.find(_.toString.equalsIgnoreCase(s))
}

/**
 * A function for testing how the elements of a collection are contained in another.
 * 
 * @param toString string representing the containment test
 * @author Alec Roelke
 */
enum Containment(override val toString: String) extends ((Iterable[?], Iterable[?]) => Boolean) {
  override def apply(a: Iterable[?] , b: Iterable[?]) = this match {
    case AllOf => val aSet = a.toSet; b.isEmpty || b.forall(aSet.contains)
    case AnyOf => val aSet = a.toSet; b.isEmpty || b.exists(aSet.contains)
    case Exactly =>
      def counts(it: Iterable[?]) = it.groupBy(identity).map{ case (x, xs) => x -> xs.size }
      counts(a) == counts(b)
    case NoneOf => !AnyOf(a, b)
    case SomeOf => !AllOf(a, b)
    case NotExactly => !Exactly(a, b)
  }

  /** The first collection contains all elements of the second. */
  case AllOf      extends Containment("contains all of")
  /** The first collection contains any element in the second. */
  case AnyOf      extends Containment("contains any of")
  /** The first collection and the second have the same elements in the same amounts. */
  case Exactly    extends Containment("contains exactly")
  /** The first collection contains no elements in the second. */
  case NoneOf     extends Containment("contains none of")
  /** The first collection contains some, but not all, of the elements in the second. */
  case SomeOf     extends Containment("contains not all of")
  /** The first collection may contian elements in the second but not exactly the same in the same amounts. */
  case NotExactly extends Containment("contains not exactly")
}